### ISSUE TITLE
Simplify class structure

### ISSUE THREAD

**Issue by averbraeck** [averbraeck/djunits#2]
The current class structure for types of `Scalar`, `Vector` and `Matrix` is tremendously complex, especially given that the code for these classes is generated. To show the complexity for, e.g., a `Length`:
- `Length` extends `AbstractDoubleScalarRelWithAbs<PositionUnit, Position, LengthUnit, Length>`
- `AbstractDoubleScalarRelWithAbs` extends `AbstractDoubleScalarRel` implements `DoubleScalarInterface.RelWithAbs`
- `AbstractDoubleScalarRel` extends `AbstractDoubleScalar` implements `DoubleScalarInterface.Rel`
- `AbstractDoubleScalar` extends `AbstractScalar` implements `DoubleScalarInterface`
- `AbstractScalar` extends `Number` implements `Scalar`
- `Scalar` extends `Value`, `Comparable`
- `Value` extends `ValueFunctions`
- `ValueFunctions` extends `Serializable`
- `DoubleScalarInterface.RelWithAbs` extends `DoubleScalarInterface`, `Scalar.RelWithAbs`
- `DoubleScalarInterface` extends `Scalar`
- `Scalar.RelWithAbs` extends `Scalar.Rel`
- `Scalar.Rel` extends `Scalar`, `Relative`
- `Relative`
- `DoubleScalarInterface.Rel`  extends `DoubleScalarInterface`, `Scalar.Rel`

In other words, 14 classes and interfaces define the contract and implementation of the `Length` scalar. And I left out the tremendously complex generics like:

```java
interface RelWithAbs<AU extends AbsoluteLinearUnit<AU, RU>, A extends DoubleScalarInterface.Abs<AU, A, RU, R>,
        RU extends Unit<RU>, R extends DoubleScalarInterface.RelWithAbs<AU, A, RU, R>>
        extends DoubleScalarInterface<RU, R>, Scalar.RelWithAbs<AU, A, RU, R>
```

With the `default` option in interfaces, and the possibility to define static functions in interfaces, several abstract classes can be removed from the hierarchy, without changing anything in the contract of a `Scalar`, `Vector` or `Matrix` such as the `Length`.

**Comment by averbraeck** [comment]
Base idea (without generics):

- `interface Value extends Serializable, Cloneable`
  defines displayUnit, absolute, relative, toString, common functions for abs and rel
- `abstract class Scalar extends Number implements Value`
  implements some of the Value methods
- `abstract class IndexedValue implements Value`
  defines the methods for StorageType, mutable, immutable, isDense, isSparse, toDense, toSparse
- `abstract class Vector extends IndexedValue implements Iterable`
  implements some of the base methods for a Vector
- `abstract class Matrix extends IndexedValue`
  implements some of the base methods for a Matrix

`Scalar`, `Vector` and `Matrix` have a `Rel`, `Abs` and `RelWithAbs` inner class to define a number of abstract methods belonging specifically to those types of quantities.

For the specific implementation for `DoubleScalar` (as an example -- the design for Vector and Matrix and for Float is similar), we now have 6 classes: `DoubleScalarInterface`, `DoubleScalar`, `AbstractDoubleScalar`, `AbstractDoubleScalarRel`, `AbstractDoubleScalarAbs`, and `AbstractDoubleScalarRelWithAbs`. This will be reduced to one interface `DoubleScalar` with three inner classes:  `Rel`, `Abs` and `RelWithAbs`  to be able to extend the `Scalar.Rel`, `Scalar.Abs` and `Scalar.RelWithAbs` abstract classes. `DoubleScalar` implements all methods using default methods -- they almost all use `getSI()`.

**Comment by averbraeck** [comment]
Now with the generics (first design):

The Value as the basis:
- `public interface Value<U extends Unit<U>, T extends Value<U, T>> extends Serializable, Cloneable`
  Every value has a unit, and has to be compatible with that unit.

The Scalar for 0-domensional Values:
- `public abstract class Scalar<U extends Unit<U>, S extends Scalar<U, S>> extends Number implements Value<U, S>, Comparable<S>`
  A scalar is a Java Number that implements the Value functions for that Scalar and Unit
- `public abstract class Scalar.Rel<U extends Unit<U>, R extends Scalar.Rel<U, R>> extends Scalar<U, R> implements Relative<U, R>`
  A relative scalar does not have an 'attached' absolute scalar.
- `public abstract class Scalar.RelWiithAbs<AU extends AbsoluteLinearUnit<AU, RU>, A extends Scalar.Abs<AU, A, RU, R>, RU extends Unit<RU>, R extends Scalar.RelWithAbs<AU, A, RU, R>> extends Scalar.Rel<RU, R>`
  The relative scalar R (e.g., Length) with unit RU has an attached absolute scalar A (e.g., Position) with unit AU. 
- `public abstract class Scalar.Abs<AU extends AbsoluteLinearUnit<AU, RU>, A extends Scalar.Abs<AU, A, RU, R>, RU extends Unit<RU>, R extends Scalar.RelWithAbs<AU, A, RU, R>> extends Scalar<AU, A> implements Absolute`
  The absolute scalar A (e.g., Position) with unit AU has an attached relative scalar R (e.g., Length) with unit RU.

The IndexedValue fpr Vectors and Matrices:
- `public abstract class IndexedValue<U extends Unit<U>, S extends Scalar<U, S>, T extends IndexedValue<U, S, T, D>, D extends Storage<D>> implements Value<U, T>`
  The indexed value defines classes (T) of a higher dimension than Scalar that do have a corresponding Scalar (S) with a Unit (U). In addition, they store their data (D) using some kind of Storage.

**Comment by averbraeck** [comment]
Based on the above, `DoubleScalar` would be defined as follows:

- `public abstract class DoubleScalar<U extends Unit<U>, S extends DoubleScalar<U, S>> extends Number implements Scalar<U, S>`
  This will define functions where the `double` value such as `double getSI()` extends what can be defined for a Scalar where the numeric definition of the quantity value is not yet specified.

The `DoubleScalar` class can be implemented as follows:
- `public abstract class DoubleScalar.Rel<U extends Unit<U>, S extends DoubleScalar.Rel<U, S>> extends DoubleScalar<U, S> implements Relative<U, S>`
- `public abstract class DoubleScalar.RelWithAbs<AU extends AbsoluteLinearUnit<AU, RU>, A extends Scalar.Abs<AU, A, RU, R>, RU extends Unit<RU>, R extends DoubleScalar.RelWithAbs<AU, A, RU, R>> extends DoubleScalar<RU, R> implements AbsWithRel<AU, A, RU, R>`
- `public abstract class DoubleScalar.Absolute<AU extends AbsoluteLinearUnit<AU, RU>, A extends Scalar.Abs<AU, A, RU, R>, RU extends Unit<RU>, R extends DoubleScalar.RelWithAbs<AU, A, RU, R>> extends DoubleScalar<AU, A> implements Absolute<AU, A, RU, R>`

**Comment by averbraeck** [comment]
This implementation would simplify the number of interfaces and classes. A `Speed` would implement/extend as follows:

```
   Value --> Serializable, Cloneable
     ^
   Scalar --> Comparable
     ^
DoubleScalar --> Number
     ^
DoubleScalar.Rel --> Relative
     ^
   Speed
```

A `Length` would implement/extend as follows:

```
   Value --> Serializable, Cloneable
     ^
   Scalar --> Comparable
     ^
DoubleScalar --> Number
     ^
DoubleScalar.AbsWithRel --> AbsWithRel --> Relative
     ^
   Speed
```

**Comment by averbraeck** [comment]
- Feature branch merged into main branch
- djunits-demo project adapted to changes
- djunits-generator project adapted to changes (template classes)
- unit tests show 100% success; 92% total coverage (less than earlier b/c of removal of a lot of 'dummy' code)
